src/PointwiseFunctions/Hydro/StressEnergy.cpp

```diff
@@ -70,23 +101,99 @@ void stress_trace(gsl::not_null<Scalar<DataType>*> result,
+    const Scalar<DataType>& lapse,
+    const Scalar<DataType>& comoving_magnetic_field_magnitude,
+    const tnsr::I<DataType, 3>& spatial_velocity,
+    const tnsr::I<DataType, 3>& shift,
+    const tnsr::I<DataType, 3>& magnetic_field,
+    const tnsr::ii<DataType, 3>& spatial_metric,
+    const tnsr::II<DataType, 3>& inverse_spatial_metric) {
+  // Preallocate to minimize number of allocations.
+  TempBuffer<tmpl::list<::Tags::TempScalar<0, DataType>,
+                        ::Tags::TempScalar<1, DataType>,
+                        ::Tags::TempScalar<2, DataType>,
+                        ::Tags::TempA<3, 3, Frame::Inertial, DataType>,
+                        ::Tags::TempA<4, 3, Frame::Inertial, DataType>,
+                        ::Tags::TempAA<5, 3, Frame::Inertial, DataType>>>
+      buffer(get_size(get(rest_mass_density)));
+
+  auto& magnetic_field_dot_spatial_velocity =
+      get<::Tags::TempScalar<0, DataType>>(buffer);
+  auto& rho_h_star = get<::Tags::TempScalar<1, DataType>>(buffer);
+  auto& p_star = get<::Tags::TempScalar<2, DataType>>(buffer);
+  auto& comoving_magnetic_field_v =
+      get<::Tags::TempA<3, 3, Frame::Inertial, DataType>>(buffer);
+  auto& four_velocity_v =
+      get<::Tags::TempA<4, 3, Frame::Inertial, DataType>>(buffer);
+  auto& inverse_spacetime_metric_v =
+      get<::Tags::TempAA<5, 3, Frame::Inertial, DataType>>(buffer);
+
+  gr::inverse_spacetime_metric(make_not_null(&inverse_spacetime_metric_v),
+                               lapse, shift, inverse_spatial_metric);
+
+  dot_product(make_not_null(&magnetic_field_dot_spatial_velocity),
+              magnetic_field, spatial_velocity, spatial_metric);
+
+  comoving_magnetic_field(make_not_null(&comoving_magnetic_field_v),
+                          spatial_velocity, magnetic_field,
+                          magnetic_field_dot_spatial_velocity, lorentz_factor,
+                          shift, lapse);
+
+  four_velocity(make_not_null(&four_velocity_v), spatial_velocity, shift,
+                lorentz_factor, lapse);
+
+  get(rho_h_star) = (get(rest_mass_density) +
+                     get(rest_mass_density) * get(specific_internal_energy)) +
+                    get(pressure) +
+                    square(get(comoving_magnetic_field_magnitude));
+
+  get(p_star) =
+      get(pressure) + 0.5 * square(get(comoving_magnetic_field_magnitude));
+
+  for (size_t i = 0; i < 4; ++i) {
+    result->get(i, i) = get(rho_h_star) * square(four_velocity_v.get(i)) +
+                        get(p_star) * inverse_spacetime_metric_v.get(i, i) -
+                        square(comoving_magnetic_field_v.get(i));
+
+    for (size_t j = i + 1; j < 4; ++j) {
+      result->get(i, j) =
+          get(rho_h_star) * four_velocity_v.get(i) * four_velocity_v.get(j) +
+          get(p_star) * inverse_spacetime_metric_v.get(i, j) -
+          comoving_magnetic_field_v.get(i) * comoving_magnetic_field_v.get(j);
+    }
```

Comment:
why do you have these two separate? Isn't it the same equation for j=i?

Reply:
done.